### DISCUSSION TITLE
Region values and annotations

### DISCUSSION THREAD

**Discussion by JonathanStarup** [flix/flix#8478] [Ideas]
- Remove regions as values:
  - They feel bad syntactically - often one trivial correct choice, and it's cumbersome to carry the region value around
  - if we rule out actual semantics to regions, like deallocation (see other issues why I think that is a good choice), the region value itself will always continue to be a dead value not used for anything
  - So the region value is a cumbersome un-inferred type application - not great.
- We add an `exp @ r` as a general expression that says that `exp,` whatever type it is, uses the region `r.`
  - You can now write `Array.empty()` or `Array.empty() @ r` if you want to.
  - This new `@` smart annotation is pretty fancy. In its base form, it could require a known lefthand side, and annotate the rightmost type variable, like `Array[_, r]` or `Ref[r]`. It can also be smarter with a new type-constraint.

**Comment by magnus-madsen** [comment]
Counter-point: The region can be used to track resources (e.g. files) and threads which must be closed explicitly upon region exit.

**Reply by JonathanStarup** [reply]
I'm thinking of something like capability passing. we can have region values without having surface region values

**Comment by JonathanStarup** [comment]
We monomorphize in regards to the region - surely that can be instrumented by the compiler if we want to.

The value is the single value of the type, it seems to me like its always redundant, unless you use it as a type anotation

**Reply by magnus-madsen** [reply]
Two values are not the same; they may have different threads and e.g. files associated with them.

**Reply by JonathanStarup** [reply]
sorry, that was unclear by me. There is only one region value per region type, so if you know the type you know the value.

**Comment by JonathanStarup** [comment]
@magnus-madsen Another point is that it would fix immutable/mutable behaviour regarding traits. Creation of collections could be abstracted over using the same signature, etc.
